src/uct/ib/ud/base/ud_ep.c

```diff
@@ -722,6 +722,31 @@ ucs_status_t uct_ud_ep_flush_nolock(uct_ud_iface_t *iface, uct_ud_ep_t *ep,
     return UCS_INPROGRESS;
 }
 
+void uct_ud_tx_wnd_purge_outstanding(uct_ud_iface_t *iface, uct_ud_ep_t *ud_ep,
+                                     ucs_status_t status)
+{
+    uct_ud_comp_desc_t *cdesc;
+    uct_ud_send_skb_t  *skb;
+
+    uct_ud_ep_tx_stop(ud_ep);
+
+    ucs_queue_for_each_extract(skb, &ud_ep->tx.window, queue, 1) {
+        skb->flags |= UCT_UD_SEND_SKB_FLAG_ERR;
+        skb->status = status;
+        if (ucs_likely(!(skb->flags & UCT_UD_SEND_SKB_FLAG_COMP))) {
+            skb->len = 0;
+        }
+        cdesc = uct_ud_comp_desc(skb);
+        /* don't call user completion from async context. instead, put
+         * it on a queue which will be progressed from main thread.
+         */
+        ucs_queue_push(&iface->tx.async_comp_q, &skb->queue);
```

Comment:
need to extract this code to new function uct_ud_iface_ass_async_comp() - also from uct_ud_ep_process_ack() and uct_ud_tx_wnd_purge_outstanding()